### PR TITLE
Allow Cripts to be used directly as @plugin

### PR DESCRIPTION
**Pull request by zwoop**
An example compiler script would be (for macOS):

```
#!/bin/sh

SOURCE=$1
DEST=$2
CXXFLAGS="-std=c++20 -I/opt/homebrew/include -I/opt/ats/include"

exec clang++ -undefined dynamic_lookup -shared -fPIC ${CXXFLAGS} -L/opt/ats/lib -lcripts -o $DEST $SOURCE
```

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci autest]

File: doc/admin-guide/files/records.yaml.en.rst [hunk 1]
```diff
@@ -4901,6 +4901,20 @@ Plug-in Configuration
    Enables (``1``) or disables (``0``) the dynamic reload feature for remap
    plugins (`remap.config`). Global plugins (`plugin.config`) do not have dynamic reload feature yet.
 
+.. ts:cv:: CONFIG proxy.config.plugin.compiler_path STRING ""
+
+   Specifies an option compiler tool path for compiling plugins. This tool should
```
Comment (bneradt) [inline doc/admin-guide/files/records.yaml.en.rst:4906]: `option` -> `optional`

File: tools/cripts/compiler.sh [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+if [ ! -f "$SOURCE" ]; then
+    echo "Source file $SOURCE does not exist"
+    exit 1
+fi
+
+if [ "${DEST_FILE}" != "${SOURCE_FILE}.so" ]; then
+    echo "Destination file name must match source file name with .so extension"
+    exit 1
+fi
```
Comment (bneradt) [inline tools/cripts/compiler.sh:66]: So if the user passes in `cripts.cc`, `DEST_FILE` should be `cripts.cc.so`? Maybe we should remove the extension from `$SOURCE_FILE` via `${SOURCE_FILE%.*}`.

File: tools/cripts/compiler.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+cache_file=""
+if [ -d "$CACHE_DIR" ]; then
+    cache_tree="${CACHE_DIR}${SOURCE_DIR}"
+    cache_file="${cache_tree}/${DEST_FILE}"
+
+    [ -d "$cache_tree" ] || mkdir -p "$cache_tree"
+
+    if [ ! "$SOURCE" -nt "$cache_file" ]; then
```
Comment (bneradt) [inline tools/cripts/compiler.sh:75]: Perhaps: `[ "$SOURCE" -ot "$cache_file" ]` (i.e., use `-ot`).

File: src/proxy/http/remap/PluginFactory.cc [hunk 4]
```diff
@@ -175,6 +183,20 @@ PluginFactory::getRemapPlugin(const fs::path &configPath, int argc, char **argv,
...
 
+      // Special case for Cripts
+      if (!runtimePath.string().ends_with(".so")) {
+        if (_compilerPath.empty()) {
+          error.assign("compiler path not set for compiling plugins");
+          return nullptr;
+        }
+
+        // Add .so to the source file, so e.g. example.cc.so. ToDo: libswoc doesn't allow appending to the extension.
+        std::string newPath = runtimePath.string();
```
Comment (bneradt) [inline src/proxy/http/remap/PluginFactory.cc:194]: Oh, I suppose we will compile it to `example.cc.so`. It shouldn't be too hard to remove the extension after the last `.` right?

**Review by bneradt — CHANGES_REQUESTED** [review]
Looks great. I just have a few suggestions to tweak things.

File: tools/cripts/compiler.sh (hunk 2 above)
Comment (zwoop) [reply tools/cripts/compiler.sh:66]: I did that intentional for a few reasons, but primarily to not make assumptions what extension (if any) people will use, and also assure consistent naming. You can now see exactly what file was compiled. 

This is sort of inline with how cmake does it as well :)

Comment (bneradt) [reply tools/cripts/compiler.sh:66]: Sounds reasonable.

File: doc/admin-guide/files/records.yaml.en.rst (hunk 1 above)
Comment (zwoop) [reply doc/admin-guide/files/records.yaml.en.rst:4906]: Fixed

File: tools/cripts/compiler.sh (hunk 3 above)
Comment (zwoop) [reply tools/cripts/compiler.sh:75]: Done

**Review by bneradt — APPROVED** [review]

**Review by bneradt — APPROVED** [review]

**Comment by cmcfarlen** [comment]
Cherry-picked to v10.0.x

**Pull request merged**